go/teams/invite_test.go

```diff
@@ -154,3 +156,422 @@ func TestKeybaseInviteMalformed(t *testing.T) {
+
+	// Seitan invite without `max_uses` or `etime`.
+	scInvite1 := makeTestSCForInviteLink()
+
+	// Seitan invite with `etime`.
+	scInvite2 := makeTestSCForInviteLink()
+	etime := keybase1.ToUnixTime(time.Now())
+	scInvite2.Etime = &etime
+
+	teamSectionForInvite := makeTestSCTeamSection(team)
+	for _, scInvite := range []SCTeamInvite{scInvite1, scInvite2} {
+		inviteID := scInvite.ID
+		teamSectionForInvite.Invites = &SCTeamInvites{
+			Readers: &[]SCTeamInvite{scInvite},
+		}
+
+		state, err := appendSigToState(t, team, nil /* state */, libkb.LinkTypeInvite,
+			teamSectionForInvite, me, nil /* merkleRoot */)
+		require.NoError(t, err)
+		require.NotNil(t, state.inner.ActiveInvites[keybase1.TeamInviteID(inviteID)])
+
+		// Try to do `used_invites` for an invite that does not have `max_uses`.
+		teamSectionCM := makeTestSCTeamSection(team)
+		teamSectionCM.Members = &SCTeamMembers{
+			Readers: &[]SCTeamMember{SCTeamMember(testUV)},
+		}
+		teamSectionCM.UsedInvites = []SCMapInviteIDUVPair{
+			{InviteID: inviteID, UV: testUV.PercentForm()},
+		}
+		_, err = appendSigToState(t, team, state, libkb.LinkTypeChangeMembership,
+			teamSectionCM, me, nil /* merkleRoot */)
+		requirePrecheckError(t, err)
+		require.Contains(t, err.Error(), "`used_invites` for an invite that did not have `max_uses`")
+	}
+}
+
+func TestTeamPlayerBadCompletedInvites(t *testing.T) {
+	// Test if `completed_invites` errors out when used on an invite that
+	// defines `max_uses`, and therefore `used_invites` should be used instead
+	// to mark the invite usage.
+
+	tc, team, me := setupTestForPrechecks(t, false /* implicitTeam */)
+	defer tc.Cleanup()
+
+	testUV := keybase1.UserVersion{Uid: libkb.UsernameToUID("t_alice_t"), EldestSeqno: 1}
+
+	teamSectionForInvite := makeTestSCTeamSection(team)
+	sectionInvite := makeTestSCForInviteLink()
+	maxUses := keybase1.TeamInviteMaxUses(10)
+	sectionInvite.MaxUses = &maxUses
+	inviteID := keybase1.TeamInviteID(sectionInvite.ID)
+	teamSectionForInvite.Invites = &SCTeamInvites{
+		Readers: &[]SCTeamInvite{sectionInvite},
+	}
+
+	state, err := appendSigToState(t, team, nil /* state */, libkb.LinkTypeInvite,
+		teamSectionForInvite, me, nil /* merkleRoot */)
+	require.NoError(t, err)
+	require.NotNil(t, state.inner.ActiveInvites[inviteID])
+
+	teamSectionCM := makeTestSCTeamSection(team)
+	teamSectionCM.Members = &SCTeamMembers{
+		Readers: &[]SCTeamMember{SCTeamMember(testUV)},
+	}
+	teamSectionCM.CompletedInvites = SCMapInviteIDToUV{
+		inviteID: testUV.PercentForm(),
+	}
+	_, err = appendSigToState(t, team, state, libkb.LinkTypeChangeMembership,
+		teamSectionCM, me, nil /* merkleRoot */)
+	requirePrecheckError(t, err)
+	require.Contains(t, err.Error(), "`completed_invites` for an invite with `max_uses`")
+	require.Contains(t, err.Error(), inviteID)
+}
+
+func TestTeamInvite64BitEtime(t *testing.T) {
+	t.Skip("unixtime fail: Not equal:  expected: 2030  actual  : 1851")
```

Comment:
nice thought though

Reply:
this comment was actually wrong (2030 -> 3020) but that was just a typo.

I added a ticket to solve that, the conversion from int64 UnixTime to time.Time seems to be messing everything up. The conversion seems to run the numbers pretty high, to go as far as to nanoseconds ???

```
func FromTime(t Time) time.Time {
	if t == 0 {
		return time.Time{}
	}
	return time.Unix(0, int64(t)*1000000)
}

func FromUnixTime(u UnixTime) time.Time {
	return FromTime(Time(u * 1000))
}

func (u UnixTime) Time() time.Time {
	return FromUnixTime(u)
}
```

`Time` in the code above is `keybase1.Time`, UnixTime is `keybase1.UnixTime`

We call `invite.etime.Time()` (`etime` is `keybase1.UnixTime`) to get `time.Time` to work on it after loading the team.

Reply:
https://play.golang.org/p/IN9bmiGrB0H

this seems to work. why are we doing this nsec thing?